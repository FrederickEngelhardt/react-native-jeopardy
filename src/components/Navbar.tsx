import React, { useState } from "react";
import styled, { withTheme } from "styled-components";
import { Theme } from "../screens/RootScreen";
import { toggleDarkMode } from "../store/themeStore";

const NavBarContainer = styled.View`
  background-color: ${props => (props.theme.darkMode ? "#000000" : "#64d3ff")}
  height: ${({ theme }) => theme.deviceHeight / 10}
  margin-top: 12px;
  padding: 12px;
  width: ${({ theme }) => theme.deviceWidth}
  justify-content: space-between;
  flex-direction: row;
`;

const AppTitleText = styled.Text`
  font-size: 28px;
  font-weight: 900;
  color: #fff;
  height: 100%;
`;

const DarkLightSwitch = styled.Switch`
  height: 100%;
  width: 50px;
`;

const TITLE = "SWAPI";

interface Props {
  theme: Theme;
}

const NavBar = (props: Props) => {
  const { darkMode } = props.theme;
  return (
    <NavBarContainer>
      <AppTitleText>{TITLE}</AppTitleText>
      <DarkLightSwitch
        onChange={() => toggleDarkMode(!darkMode)}
        value={darkMode}
      />
    </NavBarContainer>
  );
};

export default withTheme(NavBar);
