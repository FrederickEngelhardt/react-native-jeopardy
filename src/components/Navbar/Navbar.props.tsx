const navbarProps = {
  titleText: "SWAPI",
  scoreText: "score",
  user: { score: 0 }
};

export default navbarProps;
