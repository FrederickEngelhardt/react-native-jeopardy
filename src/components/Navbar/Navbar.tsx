import React, { useEffect } from "react";
import styled, { withTheme } from "styled-components";
import { Animated, StyleSheet } from "react-native";

import { Theme } from "../RootThemeProvider";
import { toggleDarkMode } from "../../store/themeStore";
import { black, lightBlue, white, yellow } from "../../constants/theming";

const NavBarContainer = styled.View`
  background-color: ${props => (props.theme.darkMode ? black : lightBlue)};
  height: ${({ theme }) => theme.deviceHeight / 10};
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 12px;
  width: ${({ theme }) => theme.deviceWidth};
  justify-content: space-between;
  align-items: center;
  flex-direction: row;
`;

const ScoreContainer = styled.View`
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;

const AppTitleText = styled.Text`
  font-size: 28px;
  font-weight: 900;
  color: ${({ theme }) => (theme.darkMode ? white : black)};
`;

const ScoreText = styled(AppTitleText)`
  font-size: 18px;
  margin-right: 12px;
`;

const DarkLightSwitch = styled.Switch.attrs(() => ({
  trackColor: { true: white, false: white },
  thumbColor: black
}))`
  width: 50px;
`;

const styles = StyleSheet.create({
  text: {
    fontWeight: "900"
  }
});

const TITLE_TEXT = "Jeopardy";
const SCORE_TEXT = "Score";

interface Props {
  user: { score: number };
  scoreText: string;
  theme: Theme;
  titleText: string;
  toggleDarkMode(boolean): void;
}

const NavBar = (props: Props) => {
  const {
    user: { score },
    scoreText,
    theme: { darkMode },
    titleText,
    toggleDarkMode
  } = props;

  const animatedScoreValue = new Animated.Value(0);

  const animateScore = () => {
    if (score > 0) {
      Animated.spring(animatedScoreValue, {
        toValue: 1
      }).start();
    }
  };

  useEffect(() => {
    animateScore();
  }, [score]);

  return (
    <NavBarContainer>
      <AppTitleText>{titleText}</AppTitleText>
      <ScoreContainer>
        <ScoreText>{scoreText}</ScoreText>
        <Animated.Text
          style={[
            {
              fontSize: animatedScoreValue.interpolate({
                inputRange: [0, 0.9, 1],
                outputRange: [24, 30, 24],
                extrapolate: "clamp"
              }),
              color: darkMode ? white : black
            },
            styles.text
          ]}
        >
          {score}
        </Animated.Text>
      </ScoreContainer>
      <DarkLightSwitch
        onChange={() => toggleDarkMode(darkMode)}
        value={darkMode}
      />
    </NavBarContainer>
  );
};

NavBar.defaultProps = {
  titleText: TITLE_TEXT,
  scoreText: SCORE_TEXT,
  toggleDarkMode: darkMode => toggleDarkMode(!darkMode)
};

export default withTheme(NavBar);
