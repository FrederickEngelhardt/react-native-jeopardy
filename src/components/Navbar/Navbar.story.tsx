import { storiesOf } from "@storybook/react-native";
import * as React from "react";
import { action } from "@storybook/addon-actions";
import { text, number, withKnobs } from "@storybook/addon-knobs";

import Navbar from "./Navbar";
import navbarProps from "./Navbar.props";

storiesOf("<Navbar />", module)
  .addDecorator(withKnobs)
  .add("default", () => (
    <Navbar
      toggleDarkMode={action("toggled Dark Mode")}
      titleText={text("Title", navbarProps.titleText)}
      user={{ score: number("score", navbarProps.user.score) }}
    />
  ));
