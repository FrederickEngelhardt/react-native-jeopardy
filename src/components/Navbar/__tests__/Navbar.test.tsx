import React from "react";
import { render } from "@testing-library/react-native";
import { ThemeProvider } from "styled-components/native";

import Navbar from "../Navbar";
import navbarProps from "../Navbar.props";

const theme = {
  darkMode: true,
  deviceHeight: 800,
  deviceWidth: 320
};

describe("<Navbar />", () => {
  it("renders correctly", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Navbar {...navbarProps} theme={theme}/>
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
