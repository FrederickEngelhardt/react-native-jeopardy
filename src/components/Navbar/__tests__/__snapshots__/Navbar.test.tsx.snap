// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#000",
          "flexDirection": "row",
          "height": 80,
          "justifyContent": "space-between",
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 12,
          "width": 320,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#fff",
            "fontSize": 28,
            "fontWeight": "900",
          },
        ]
      }
    >
      SWAPI
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "900",
              "marginRight": 12,
            },
          ]
        }
      >
        score
      </Text>
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
            "fontWeight": "900",
          }
        }
      >
        0
      </Text>
    </View>
    <Switch
      style={
        Array [
          Object {
            "width": 50,
          },
        ]
      }
      thumbColor="#000"
      trackColor={
        Object {
          "false": "#fff",
          "true": "#fff",
        }
      }
      value={true}
    />
  </View>
</View>
`;
